index.bs

```diff
@@ -677,14 +692,515 @@ id=] must be the same as the [=window handle=].
+object, specified as <code>RemoteObject</code>. The object type is specified in
+the <code>type</code> property. In the case of JSON-representable primitive
+objects, this contains the object value in the <code>value</code> property; in
+the case of non-JSON-representable primitives, the <code>value</code> property
+contains a string representation of the object. For non-primitive objects, the
+<code>objectId</code> property contains a string id that provides a unique
+handle to the object, valid for its lifetime inside the engine. In this case the
+<code>value</code> property can contain a complex object (either a list or a
+struct) itself containing further remote object values. This allows objects like
+arrays to have a single representation.
+
+DOM Nodes are also represented by <code>RemoteObject</code> instances. These
+have a partial serialization of the node in the value property.
+
+Note: mirror objects do not keep the original object alive in the runtime. If an
+object is discarded in the runtime subsequent attempts to access it via the
+protocol will result in an error.
+
+A [=/session=] has an <dfn>object id map</dfn>. This is a weak map from objects to
+their corresponding id.
+
+Issue: Should this be explicitly per realm?
+
+<div algorithm>
+To get the <dfn export>object id for an object</dfn> given an |object|:
+
+  1. If [=object id map=] for the [=current session=] does not contain |object|
+     run the following steps:
+
+     1. Let |object id| be a new, unique, string identifier for |object|. If
+        |object| is an [=/element=] this must be the [=web element reference=]
+        for |object|; if it's [=WindowProxy=] object, this must be the
+        [=window handle=] for |object|.
+
+    1. Set the value of |object| in the [=object id map=] to |object id|.
+
+  1. Return the result of getting the value for |object| in [=object id map=].
+</div>
+
+<div algorithm>
+To get the <dfn export>object for an object id</dfn> given an |object id|:
+
+  1. For each |object| → |id value| of [=object id map=]:
+
+    1. If |id value| is equal to |object id|, return [=success=] with data
+       |object|
+
+  1. Return [=error=] with [=error code=] [=no such element=]
+
+  Issue: This error code isn't right.
+</div>
+
+[=remote end definition=] and [=local end definition=]
+```
+RemoteObject = {
+  UndefinedObject //
+  NullObject //
+  StringObject //
+  NumberObject //
+  BooleanObject //
+  BigIntObject //
+  SymbolObject //
+  ArrayObject //
+  FunctionObject //
+  ObjectObject //
+  FunctionObject //
+  RegexObject //
+  DateObject //
+  MapObject //
+  SetObject //
+  WeakMapObject //
+  WeakSetObject //
+  IteratorObject //
+  GeneratorObject //
+  ErrorObject //
+  ProxyObject //
+  PromiseObject //
+  TypedArrayObject //
+  ArrayBufferObject //
+  NodeObject //
+  WindowProxyObject //
+}
+
+ObjectId = text;
+
+UndefinedObject = {
+  type: "undefined",
+  value: null,
```

Comment:
Including the value for undefined and null seems unnecessary, and in the case of undefined makes it easier to conflate null and undefined. I see that https://tools.ietf.org/html/rfc8610#appendix-D does have `undefined` although I don't understand what it means.

How about just no value?